dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/DefaultIconService.java

```diff
@@ -67,33 +71,48 @@ public class DefaultIconService implements IconService {
   private static final String ICON_PATH = "SVGs";
 
   private final IconStore iconStore;
-
   private final FileResourceService fileResourceService;
-
   private final UserService userService;
 
+  private final Set<DefaultIcon> ignoredAfterFailure = ConcurrentHashMap.newKeySet();
+
   @Override
-  @Transactional
-  public void createDefaultIcon(DefaultIcon icon) {
-    if (iconStore.containsKeys(getKeysForVariants(icon))) return;
-
-    for (Icon i : createIconWithVariants(icon)) {
-      if (!iconStore.containsKeys(Set.of(i.getKey()))) {
-        String fileResourceId = CodeGenerator.generateUid();
-        Resource resource = getDefaultIconResource(i.getKey());
-        try {
-          FileResource fileResource =
-              fileResourceService.syncSaveFileResource(
-                  fileResourceId, MediaType.IMAGE_PNG, resource.getInputStream(), CUSTOM_ICON);
-
-          i.setFileResource(fileResource);
-
-          addIconInternal(i);
-        } catch (Exception ex) {
-          throw new RuntimeException(ex);
+  @Transactional(readOnly = true)
+  public List<Icon> findNonExistingDefaultIcons() {
+    Set<String> existing = Set.copyOf(iconStore.getAllKeys());
+    List<Icon> missing = new ArrayList<>();
+    for (DefaultIcon icon : DefaultIcon.values()) {
+      if (!ignoredAfterFailure.contains(icon)
+          && icon.getVariants().stream().anyMatch(key -> !existing.contains(key))) {
+        for (Icon i : icon.createIcons()) {
+          if (!existing.contains(i.getKey())) {
+            missing.add(i);
+          }
         }
       }
     }
+    return missing;
+  }
+
+  @Override
+  @Transactional
+  public String uploadDefaultIcon(Icon icon) throws ConflictException {
+    String fileResourceId = CodeGenerator.generateUid();
+    Resource resource = getDefaultIconResource(icon.getKey());
+    try {
+      FileResource fileResource =
+          FileResource.ofKey(CUSTOM_ICON, icon.getKey(), MediaType.IMAGE_PNG);
+      fileResource.setUid(fileResourceId);
+      fileResource.setAssigned(true);
+      try (InputStream image = resource.getInputStream()) {
+        fileResourceService.syncSaveFileResource(fileResource, image);
+      }
+      icon.setFileResource(fileResource);
+      return fileResourceId;
+    } catch (IOException ex) {
+      ignoredAfterFailure.add(icon.getOrigin());
```

Comment:
Why are we ignoring these and not retrying later?

Reply:
How would a user notice this error and potentially fix it?

Reply:
Because it is likely this keeps failing for the same reason causing an infinite sequence of tries.
This is a very simple way to avoid that.